Bug with forward index storage (signed/unsigned problem?)

The integrated forward index looks at the number of unique terms, and uses only as many bytes as needed to store each term index in a document (the external forward index always used 4 bytes per term in a document).

It seems like there's a bug with this, specifically when 3 bytes are used. The stacktrace below has an ArrayIndexOutOfBoundsException where the incorrect index is 16777215, which is 2^24-1, or the highest value an unsigned 3-byte integer can take, or -1 for a signed 3-byte integer.

There's probably a problem with the (de)serialization of a Java (4-byte) integer to a 3-byte value.

```
Caused by: java.util.concurrent.ExecutionException: nl.inl.blacklab.exceptions.BlackLabRuntimeException: java.lang.ArrayIndexOutOfBoundsException: Index 16777215 out of bounds for length 771577
	at nl.inl.blacklab.server.search.BlsCacheEntry.get(BlsCacheEntry.java:274)
	at nl.inl.blacklab.server.search.BlsCacheEntry.get(BlsCacheEntry.java:254)
	at nl.inl.blacklab.server.search.BlsCacheEntry.get(BlsCacheEntry.java:23)
	at nl.inl.blacklab.search.results.ResultsStatsDelegate.stats(ResultsStatsDelegate.java:47)
	... 51 more
Caused by: nl.inl.blacklab.exceptions.BlackLabRuntimeException: java.lang.ArrayIndexOutOfBoundsException: Index 16777215 out of bounds for length 771577
	at nl.inl.blacklab.search.results.HitsFromQuery.ensureResultsRead(HitsFromQuery.java:219)
	at nl.inl.blacklab.search.results.ResultsAbstract.ensureAllResultsRead(ResultsAbstract.java:259)
	at nl.inl.blacklab.search.results.HitsFromQuery.hitsProcessedTotal(HitsFromQuery.java:285)
	at nl.inl.blacklab.search.results.HitsFromQuery.resultsProcessedTotal(HitsFromQuery.java:306)
	at nl.inl.blacklab.search.results.ResultsAbstract$1.processedTotal(ResultsAbstract.java:105)
	at nl.inl.blacklab.search.results.ResultCount.processedTotal(ResultCount.java:82)
	at nl.inl.blacklab.searches.SearchCountFromResults.executeInternal(SearchCountFromResults.java:46)
	at nl.inl.blacklab.searches.SearchCountFromResults.executeInternal(SearchCountFromResults.java:17)
	at nl.inl.blacklab.server.search.BlsCacheEntry.executeSearch(BlsCacheEntry.java:137)
	at nl.inl.blacklab.server.search.BlsCacheEntry.lambda$start$0(BlsCacheEntry.java:124)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	... 1 more
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 16777215 out of bounds for length 771577
	at nl.inl.blacklab.forwardindex.TermsIntegrated.segmentIdsToGlobalIds(TermsIntegrated.java:227)
	at nl.inl.blacklab.codec.BLTerms.termsEqual(BLTerms.java:117)
	at nl.inl.blacklab.search.fimatch.ForwardIndexAccessorIntegrated$ForwardIndexAccessorLeafReaderIntegrated.segmentTermsEqual(ForwardIndexAccessorIntegrated.java:105)
	at nl.inl.blacklab.search.fimatch.ForwardIndexDocumentImpl.segmentTermsEqual(ForwardIndexDocumentImpl.java:153)
	at nl.inl.blacklab.search.matchfilter.MatchFilterSameTokens.evaluate(MatchFilterSameTokens.java:96)
	at nl.inl.blacklab.search.lucene.SpansConstrained.ensureValidHit(SpansConstrained.java:149)
	at nl.inl.blacklab.search.lucene.SpansConstrained.nextStartPosition(SpansConstrained.java:118)
	at nl.inl.blacklab.search.results.SpansReader.advanceSpansToNextHit(SpansReader.java:191)
	at nl.inl.blacklab.search.results.SpansReader.run(SpansReader.java:285)
	at java.base/java.util.ArrayList.forEach(Unknown Source)
	at nl.inl.blacklab.search.results.HitsFromQuery.lambda$ensureResultsRead$3(HitsFromQuery.java:193)
	... 5 more\n'
```

--- a/engine/src/main/java/nl/inl/blacklab/codec/BlackLab40PostingsWriter.java
+++ b/engine/src/main/java/nl/inl/blacklab/codec/BlackLab40PostingsWriter.java
@@ -442,14 +442,15 @@ public class BlackLab40PostingsWriter extends FieldsConsumer {
      * @throws IOException       When failing to write
      */
     private void writeTokensInDoc(IndexOutput outTokensIndexFile, IndexOutput outTokensFile, int[] tokensInDoc) throws IOException {
-        int max = 0;
+        int max = 0, min = 0;
         boolean allTheSame = tokensInDoc.length > 0; // if no tokens, then not all the same.
         int last = -1;
         for (int token: tokensInDoc) {
             max = Math.max(max, token);
+            min = Math.min(min, token);
             allTheSame = allTheSame && (last == -1 || last == token);
             last = token;
-            if (max > Short.MAX_VALUE && !allTheSame) // stop if already at worst case (int per token + not all the same).
+            if ((min < Short.MIN_VALUE || max > Short.MAX_VALUE) && !allTheSame) // stop if already at worst case (int per token + not all the same).
                 break;
         }
 
@@ -461,9 +462,9 @@ public class BlackLab40PostingsWriter extends FieldsConsumer {
         switch (tokensCodec) {
             case ALL_TOKENS_THE_SAME: tokensCodecParameter = 0; break;
             case VALUE_PER_TOKEN: {
-                if (max <= Byte.MAX_VALUE) tokensCodecParameter = VALUE_PER_TOKEN_PARAMETER.BYTE.code;
-                else if (max <= Short.MAX_VALUE) tokensCodecParameter = VALUE_PER_TOKEN_PARAMETER.SHORT.code;
-                else if (max <= 0xFFFFFF) tokensCodecParameter = VALUE_PER_TOKEN_PARAMETER.THREE_BYTES.code;
+                if (min >= Byte.MIN_VALUE && max <= Byte.MAX_VALUE) tokensCodecParameter = VALUE_PER_TOKEN_PARAMETER.BYTE.code;
+                else if (min >= Short.MIN_VALUE && max <= Short.MAX_VALUE) tokensCodecParameter = VALUE_PER_TOKEN_PARAMETER.SHORT.code;
+                else if (min >= ThreeByteInt.MIN_VALUE && max <= ThreeByteInt.MAX_VALUE) tokensCodecParameter = VALUE_PER_TOKEN_PARAMETER.THREE_BYTES.code;
                 else tokensCodecParameter = VALUE_PER_TOKEN_PARAMETER.INT.code;
                 break;
             }
@@ -496,9 +497,7 @@ public class BlackLab40PostingsWriter extends FieldsConsumer {
                     break;
                 case THREE_BYTES:
                     for (int token : tokensInDoc) {
-                        outTokensFile.writeByte((byte)(token >> 16));
-                        outTokensFile.writeByte((byte)(token >>  8));
-                        outTokensFile.writeByte((byte) token);
+                        ThreeByteInt.write((b) -> outTokensFile.writeByte(b), token);
                     }
                     break;
                 case INT:

--- a/engine/src/main/java/nl/inl/blacklab/codec/SegmentForwardIndex.java
+++ b/engine/src/main/java/nl/inl/blacklab/codec/SegmentForwardIndex.java
@@ -174,8 +174,7 @@ class SegmentForwardIndex implements AutoCloseable {
                         case THREE_BYTES:
                             _tokens.seek(docTokensOffset + (long) start * 3);
                             for (int j = 0; j < snippet.length; j++) {
-                                snippet[j] = (((_tokens.readByte() & 0xFF) << 16)
-                                        | ((_tokens.readByte() & 0xFF) <<  8) |  (_tokens.readByte() & 0xFF));
+                                snippet[j] = ThreeByteInt.read( () -> _tokens.readByte() );
 
                             }
                             break;

--- a/engine/src/main/java/nl/inl/blacklab/codec/ThreeByteInt.java
+++ b/engine/src/main/java/nl/inl/blacklab/codec/ThreeByteInt.java
@@ -1,0 +1,49 @@
+package nl.inl.blacklab.codec;
+
+import java.io.IOException;
+
+/**
+ * Provides a way to serialize a 3-byte signed integer.
+ *
+ * Careful: the write() method does NOT check if the given
+ * integer is within MIN_VALUE and MAX_VALUE. You should make sure
+ * of this.
+ */
+class ThreeByteInt {
+
+    @FunctionalInterface
+    interface ByteOutput {
+        void writeByte(byte b) throws IOException;
+    }
+
+    @FunctionalInterface
+    interface ByteInput {
+        byte readByte() throws IOException;
+    }
+
+    /**
+     * Minimum value that can be stored in a three-byte int (-2^23).
+     */
+    public static int MIN_VALUE = -8388608;
+
+    /**
+     * Maximum value that can be stored in a three-byte int (2^23-1).
+     */
+    public static int MAX_VALUE = 8388607;
+
+    public static void write(ByteOutput out, int value) throws IOException {
+        out.writeByte((byte) (value >> 16));
+        out.writeByte((byte) (value >> 8));
+        out.writeByte((byte) value);
+    }
+
+    public static int read(ByteInput in) throws IOException {
+        int v = (((in.readByte() & 0xFF) << 16) | ((in.readByte() & 0xFF) << 8) | (in.readByte() & 0xFF));
+        // Is the three-byte integer's sign bit set?
+        if ((v & 0x800000) != 0) {
+            // Yes. Make the most significant byte 0xFF to get the correct 4-byte integer.
+            v |= 0xFF000000;
+        }
+        return v;
+    }
+}

--- a/engine/src/test/java/nl/inl/blacklab/codec/TestThreeByteInt.java
+++ b/engine/src/test/java/nl/inl/blacklab/codec/TestThreeByteInt.java
@@ -1,0 +1,22 @@
+package nl.inl.blacklab.codec;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+public class TestThreeByteInt {
+
+    @Test
+    public void testAllValues() throws IOException {
+        ByteBuffer buf = ByteBuffer.allocate(3);
+        for (int i = ThreeByteInt.MIN_VALUE; i <= ThreeByteInt.MAX_VALUE; i++) {
+            buf.rewind();
+            ThreeByteInt.write(b -> buf.put(b), i);
+            buf.rewind();
+            int j = ThreeByteInt.read(() -> buf.get());
+            Assert.assertEquals(j, i);
+        }
+    }
+}

